2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,493 @@
+## Assumptions/simplifications
+
+- There is a single token type
+
+## Data types
+
+```go
+type Addr
+type Key
+type Epoch uint
+type VotingPower uint
+
+type Validator struct {
+  consensus_key map<Epoch, Key>
+  state map<Epoch, {inactive, candidate, frozen, jailed}>
```

Comment:
I think the `frozen` state may need to be a separate field as it needs to be set as soon as evidence is found

Reply:
That's the behavior now, but I agree this needs more analysis as it may interfere with other behaviors.